File: src/api/decoders/assessment_student_view.ts
// Strict decoding for the answer-free, no-write Instructor Student View.

import type { InstructorStudentView } from "../../../generated/api/InstructorStudentView";
import type { InstructorStudentViewDelivery } from "../../../generated/api/InstructorStudentViewDelivery";
import type { InstructorStudentViewEntry } from "../../../generated/api/InstructorStudentViewEntry";
import type { InstructorStudentViewQuestion } from "../../../generated/api/InstructorStudentViewQuestion";
import { MAX_ASSESSMENT_ORDERED_ENTRIES } from "../../../generated/api/MAX_ASSESSMENT_ORDERED_ENTRIES";
import { MAX_QUESTION_POOL_ITEMS_PER_ASSESSMENT_ENTRY } from "../../../generated/api/MAX_QUESTION_POOL_ITEMS_PER_ASSESSMENT_ENTRY";
import {
  DecodeError,
  decodeNonnegativeInteger,
  decodeNullable,
  decodePositiveInteger,
  decodeRecord,
  decodeString,
  decodeStringEnum,
} from "../decoder";
import {
  decodeAssessmentTitle,
  decodeBoundedArray,
  decodeQuestionRevisionTuple,
  decodeTimestamp,
  field,
  requireOnlyFields,
} from "./shared";
import { decodeAccountTimeZone } from "./student_assessment_decision";

const ASSESSMENT_STATUSES = ["unreleased", "released", "closed", "archived"] as const;
const LATE_WORK_RULES = ["accept", "mark_late", "reject"] as const;

function decodeAssessmentEditNumber(value: unknown, path: string): string {
  const assessmentEditNumber = decodeString(value, path);
  if (
    !/^[1-9][0-9]*$/u.test(assessmentEditNumber) ||
    BigInt(assessmentEditNumber) > 9_223_372_036_854_775_807n
  ) {
    throw new DecodeError(path, "a positive Assessment Edit Number");
  }
  return assessmentEditNumber;
}

function decodeInstructions(value: unknown, path: string): string {
  const instructions = decodeString(value, path);
  if (instructions.includes("\0") || Array.from(instructions).length > 50_000) {
    throw new DecodeError(path, "plain-text instructions no longer than 50000 Unicode scalars");
  }
  return instructions;
}

function decodePolicyLimit(value: unknown, path: string): number | null {
  return decodeNullable(value, path, (entry, entryPath) => {
    const limit = decodePositiveInteger(entry, entryPath);
    if (limit > 2_147_483_647) throw new DecodeError(entryPath, "a bounded positive integer");
    return limit;
  });
}

function decodeDelivery(value: unknown, path: string): InstructorStudentViewDelivery {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "available_at",
    "due_at",
    "closes_at",
    "assessment_attempt_time_limit_seconds",
    "attempt_limit",
    "late_work_rule",
  ]);
  return {
    available_at: decodeNullable(
      field(record, "available_at", path),
      `${path}.available_at`,
      decodeTimestamp,
    ),
    due_at: decodeNullable(field(record, "due_at", path), `${path}.due_at`, decodeTimestamp),
    closes_at: decodeNullable(
      field(record, "closes_at", path),
      `${path}.closes_at`,
      decodeTimestamp,
    ),
    assessment_attempt_time_limit_seconds: decodePolicyLimit(
      field(record, "assessment_attempt_time_limit_seconds", path),
      `${path}.assessment_attempt_time_limit_seconds`,
    ),
    attempt_limit: decodePolicyLimit(field(record, "attempt_limit", path), `${path}.attempt_limit`),
    late_work_rule: decodeStringEnum(
      field(record, "late_work_rule", path),
      `${path}.late_work_rule`,
      LATE_WORK_RULES,
    ),
  };
}

function decodeInstructorStudentViewQuestion(
  value: unknown,
  path: string,
): InstructorStudentViewQuestion {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, ["position", "questionRevisionTuple"]);
  return {
    position: decodePositiveInteger(field(record, "position", path), `${path}.position`),
    questionRevisionTuple: decodeQuestionRevisionTuple(
      field(record, "questionRevisionTuple", path),
      `${path}.questionRevisionTuple`,
      true,
    ),
  };
}

function decodeEntry(value: unknown, path: string): InstructorStudentViewEntry {
  const record = decodeRecord(value, path);
  const entryKind = decodeString(field(record, "kind", path), `${path}.kind`);
  const authoredPosition = decodeNonnegativeInteger(
    field(record, "authoredPosition", path),
    `${path}.authoredPosition`,
  );
  if (entryKind === "presented") {
    requireOnlyFields(record, path, ["kind", "authoredPosition", "questions"]);
    const questions = decodeBoundedArray(
      field(record, "questions", path),
      `${path}.questions`,
      MAX_QUESTION_POOL_ITEMS_PER_ASSESSMENT_ENTRY,
      decodeInstructorStudentViewQuestion,
    );
    if (questions.length === 0) {
      throw new DecodeError(`${path}.questions`, "one or more presented Questions");
    }
    return { kind: "presented", authoredPosition, questions };
  }
  if (entryKind === "notShown") {
    requireOnlyFields(record, path, ["kind", "authoredPosition", "reason"]);
    return {
      kind: "notShown",
      authoredPosition,
      reason: decodeStringEnum(field(record, "reason", path), `${path}.reason`, [
        "assessmentEntryUnavailable",
      ] as const),
    };
  }
  throw new DecodeError(`${path}.kind`, "presented or notShown");
}

/** Decode one closed manifest and enforce both authored and flattened order. */
export function decodeInstructorStudentView(
  value: unknown,
  path = "response",
): InstructorStudentView {
  const record = decodeRecord(value, path);
  requireOnlyFields(record, path, [
    "assessmentEditNumber",
    "status",
    "title",
    "instructions",
    "displayTimeZone",
    "delivery",
    "entries",
  ]);
  const entries = decodeBoundedArray(
    field(record, "entries", path),
    `${path}.entries`,
    MAX_ASSESSMENT_ORDERED_ENTRIES,
    decodeEntry,
  );
  const questionPositions: number[] = [];
  entries.forEach((entry, index) => {
    if (entry.authoredPosition !== index) {
      throw new DecodeError(
        `${path}.entries[${index}].authoredPosition`,
        `the ordered position ${index}`,
      );
    }
    if (entry.kind === "presented") {
      questionPositions.push(...entry.questions.map((question) => question.position));
    }
  });
  const orderedPositions = [...questionPositions].sort((left, right) => left - right);
  orderedPositions.forEach((position, index) => {
    if (position !== index + 1) {
      throw new DecodeError(
        `${path}.entries`,
        "unique flattened Question positions numbered consecutively from 1",
      );
    }
  });
  return {
    assessmentEditNumber: decodeAssessmentEditNumber(
      field(record, "assessmentEditNumber", path),
      `${path}.assessmentEditNumber`,
    ),
    status: decodeStringEnum(field(record, "status", path), `${path}.status`, ASSESSMENT_STATUSES),
    title: decodeAssessmentTitle(field(record, "title", path), `${path}.title`),
    instructions: decodeInstructions(field(record, "instructions", path), `${path}.instructions`),
    displayTimeZone: decodeAccountTimeZone(
      field(record, "displayTimeZone", path),
      `${path}.displayTimeZone`,
    ),
    delivery: decodeDelivery(field(record, "delivery", path), `${path}.delivery`),
    entries,
  };
}

File: crates/question_model/src/assessment_student_view.rs
//! Browser-safe contracts for the no-write Instructor Student View.
//!
//! The initial response is a bounded manifest rather than an eager rendering
//! of every Question. A separate one-Question read reauthorizes the current
//! Assessment Edit Number and reuses the ordinary answer-free Student Question
//! presentation contract.

use std::num::NonZeroU32;

use serde::Serialize;

use crate::{
    AccountTimeZone, AssessmentEditNumber, AssessmentInstructions, AssessmentStatus,
    AssessmentTitle, LateWorkRule, QuestionRevisionTuple, Timestamp,
};

/// Answer-free, non-mutating Instructor Student View manifest for one current Assessment.
///
/// ASVS 8.1.1, 8.1.2, and 8.3.1: the trusted server admits this projection only
/// through the authenticated Instructor's direct Course relationship. The DTO
/// exposes neither a write operation nor Student-specific fields.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InstructorStudentView {
    /// Exact current authored Assessment state used for this projection.
    pub assessment_edit_number: AssessmentEditNumber,
    /// Current stable Assessment lifecycle.
    pub status: AssessmentStatus,
    /// Student-facing Assessment title.
    pub title: AssessmentTitle,
    /// Student-facing instructions.
    pub instructions: AssessmentInstructions,
    /// Authenticated Instructor's IANA zone for presenting delivery facts.
    pub display_time_zone: AccountTimeZone,
    /// Server-derived base delivery facts, without Student progress or actions.
    pub delivery: InstructorStudentViewDelivery,
    /// Current entries in authored order, preserving unavailable positions.
    pub entries: Vec<InstructorStudentViewEntry>,
}

/// Instructor-base delivery facts shown as Student-facing Assessment policy.
///
/// These facts describe Assessment policy, never a particular Student's state.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InstructorStudentViewDelivery {
    #[serde(rename = "available_at")]
    pub available_at: Option<Timestamp>,
    #[serde(rename = "due_at")]
    pub due_at: Option<Timestamp>,
    #[serde(rename = "closes_at")]
    pub closes_at: Option<Timestamp>,
    #[serde(rename = "assessment_attempt_time_limit_seconds")]
    pub assessment_attempt_time_limit_seconds: Option<u32>,
    #[serde(rename = "attempt_limit")]
    pub attempt_limit: Option<u32>,
    #[serde(rename = "late_work_rule")]
    pub late_work_rule: LateWorkRule,
}

/// One current Assessment Entry in the ordered Student View manifest.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(
    tag = "kind",
    rename_all = "camelCase",
    rename_all_fields = "camelCase"
)]
pub enum InstructorStudentViewEntry {
    /// One fixed Question or the selected Questions from one Question Pool.
    Presented {
        /// Zero-based position in the current authored Assessment Entry order.
        authored_position: u32,
        /// Presented Questions in their projected preview order.
        questions: Vec<InstructorStudentViewQuestion>,
    },
    /// One current authored entry excluded from Student delivery.
    NotShown {
        /// Zero-based position in the current authored Assessment Entry order.
        authored_position: u32,
        reason: InstructorStudentViewNotShownReason,
    },
}

/// One presented Question in Instructor Student View: position plus its Question Revision Tuple.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct InstructorStudentViewQuestion {
    /// One-based position in the flattened presented-Question sequence.
    pub position: NonZeroU32,
    pub question_revision_tuple: QuestionRevisionTuple,
}

/// Closed Student-safe reason that an authored entry is not shown.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub enum InstructorStudentViewNotShownReason {
    /// The current Assessment Entry is unavailable for delivery.
    AssessmentEntryUnavailable,
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{QuestionId, QuestionRevisionNumber};

    fn question_revision_tuple() -> QuestionRevisionTuple {
        QuestionRevisionTuple {
            question_id: "0000-T00N".parse::<QuestionId>().expect("Question ID"),
            revision_number: QuestionRevisionNumber::new(1).expect("Question Revision"),
        }
    }

    #[test]
    fn presented_entry_serializes_only_public_exact_question_locators() {
        let question_id = "0000-T00N".parse::<QuestionId>().expect("Question ID");
        let entry = InstructorStudentViewEntry::Presented {
            authored_position: 0,
            questions: vec![InstructorStudentViewQuestion {
                position: NonZeroU32::new(1).expect("positive position"),
                question_revision_tuple: question_revision_tuple(),
            }],
        };
        assert_eq!(
            serde_json::to_value(entry).expect("manifest entry serializes"),
            serde_json::json!({
                "kind": "presented",
                "authoredPosition": 0,
                "questions": [{
                    "position": 1,
                    "questionRevisionTuple": {
                        "questionId": question_id.to_string(),
                        "revisionNumber": 1
                    }
                }]
            })
        );
    }

    #[test]
    fn instructor_student_view_serializes_assessment_edit_number() {
        let view = InstructorStudentView {
            assessment_edit_number: "3".parse().expect("Assessment Edit Number"),
            status: AssessmentStatus::Unreleased,
            title: crate::AssessmentTitle::try_new("Quiz".to_string()).expect("title"),
            instructions: AssessmentInstructions::default(),
            display_time_zone: AccountTimeZone::parse("America/Chicago").expect("zone"),
            delivery: InstructorStudentViewDelivery {
                available_at: None,
                due_at: None,
                closes_at: None,
                assessment_attempt_time_limit_seconds: None,
                attempt_limit: None,
                late_work_rule: LateWorkRule::Reject,
            },
            entries: Vec::new(),
        };
        let wire = serde_json::to_value(&view).expect("manifest serializes");
        assert_eq!(wire["assessmentEditNumber"], "3");
        assert!(wire.get("editNumber").is_none());
    }
}
